imap/caldav_alarm.c

```diff
@@ -865,20 +838,169 @@ static void process_one_record(struct mailbox *mailbox, uint32_t imap_uid,
+    buf_remove(&buf, 0, p - buf_base(&buf));
+
+    /* Open the SMTP connection */
+    smtpclient_t *sm = NULL;
+    r = smtpclient_open(&sm);
+    if (r) {
+        syslog(LOG_ERR, "smtpclient_open failed: %s", error_message(r));
+        goto done;
+    }
+    smtpclient_set_auth(sm, json_string_value(identity));
+
+    /* Prepare envelope */
+    smtp_envelope_t smtpenv = SMTP_ENVELOPE_INITIALIZER;
+    jmap_emailsubmission_envelope_to_smtp(&smtpenv, envelope);
+
+    /* Send message */
+    r = smtpclient_send(sm, &smtpenv, &buf);
+    if (r) {
+        syslog(LOG_ERR, "smtpclient_send failed: %s", error_message(r));
+        goto done;
+    }
+    smtp_envelope_fini(&smtpenv);
+    smtpclient_close(&sm);
+
+    /* Expunge the resource */
+    record->internal_flags |= FLAG_INTERNAL_EXPUNGED;
+    r = mailbox_rewrite_index_record(mailbox, record);
+    if (r) {
+        syslog(LOG_ERR, "expunging record (%s:%u) failed: %s",
+               mailbox->name, record->uid, error_message(r));
+    }
+
+  done:
+    if (r) {
+        /* XXX  Error handling: retry, DSN, both, neither? */
+        caldav_alarm_delete_record(mailbox->name, record->uid);
+    }
+
+    if (submission) json_decref(submission);
+    if (m) message_unref(&m);
+    buf_free(&buf);
+
+    return;
+}
+
+static void process_one_record(struct mailbox *mailbox, uint32_t imap_uid,
+                               icaltimezone *floatingtz, time_t runtime)
+{
+    int rc;
+
+    syslog(LOG_DEBUG, "processing alarms for mailbox %s uid %u",
+           mailbox->name, imap_uid);
+
+    struct index_record record;
+    memset(&record, 0, sizeof(struct index_record));
+    rc = mailbox_find_index_record(mailbox, imap_uid, &record);
+    if (rc == IMAP_NOTFOUND) {
+        syslog(LOG_ERR, "not found mailbox %s uid %u",
+               mailbox->name, imap_uid);
+        /* no record, no worries */
+        caldav_alarm_delete_record(mailbox->name, imap_uid);
+        return;
+    }
+    if (rc) {
+        syslog(LOG_ERR, "error reading mailbox %s uid %u (%s)",
+               mailbox->name, imap_uid, error_message(rc));
+        /* XXX no index record? item deleted or transient error? */
+        caldav_alarm_delete_record(mailbox->name, imap_uid);
+        return;
+    }
+    if (record.internal_flags & FLAG_INTERNAL_EXPUNGED) {
+        syslog(LOG_ERR, "already expunged mailbox %s uid %u",
+               mailbox->name, imap_uid);
+        /* no longer exists?  nothing to do */
+        caldav_alarm_delete_record(mailbox->name, imap_uid);
+        return;
+    }
+
+    if (mailbox->mbtype == MBTYPE_CALENDAR) {
+        process_valarms(mailbox, &record, floatingtz, runtime);
+    }
+#ifdef WITH_JMAP
+    else if (mailbox->mbtype == MBTYPE_SUBMISSION) {
+        process_futurerelease(mailbox, &record);
+    }
+    else if (mailbox->i.options & OPT_IMAP_HAS_ALARMS) {
+        /* XXX  Check special-use flag on mailbox */
```

Comment:
We should probably at least syslog here "something odd happening!"

Also if we don't clear the alarm, we're going to infinite loop on this I guess?  We need to make sure we always make some progress.